File: src/readers/cache.ts
import { mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from 'node:fs';
import { stat } from 'node:fs/promises';
import { join } from 'node:path';

import { isFiniteNumber, isRecord } from '../data/guards.js';
import type { DayEntry, DayMap, TokenCounts } from '../data/types.js';
import { CACHE_DIR } from '../paths.js';
import { packageVersion } from '../version.js';

/**
 * Bump when the cached shape changes. Entries written by another format are
 * dropped wholesale rather than migrated — they are rebuildable from the logs.
 */
const CACHE_FORMAT = 1;

/** One transcript file's contribution, as cached. */
export interface CachedParse {
  days: DayMap;
  /**
   * The dedup keys this file holds. A cached file still has to be checked
   * against the keys already counted, because a resumed session can copy
   * another transcript's messages into itself.
   */
  keys: string[];
}

interface CacheEntry {
  mtimeMs: number;
  size: number;
  days: Record<string, DayEntry>;
  keys: string[];
}

function isTokenCounts(value: unknown): value is TokenCounts {
  if (!isRecord(value)) return false;
  if (!isFiniteNumber(value.inputTokens) || !isFiniteNumber(value.outputTokens)) return false;
  return (
    ['cachedInputTokens', 'rawInputTokens', 'cacheCreationInputTokens', 'costUSD'] as const
  ).every((field) => value[field] === undefined || isFiniteNumber(value[field]));
}

function isDayEntry(value: unknown): value is DayEntry {
  if (!isTokenCounts(value) || !isRecord(value) || !isRecord(value.byModel)) return false;
  return Object.values(value.byModel).every((counts) => isTokenCounts(counts));
}

function isCacheEntry(value: unknown): value is CacheEntry {
  if (!isRecord(value)) return false;
  if (!isFiniteNumber(value.mtimeMs) || !isFiniteNumber(value.size)) return false;
  if (!Array.isArray(value.keys) || !value.keys.every((key) => typeof key === 'string')) {
    return false;
  }
  return isRecord(value.days) && Object.values(value.days).every((day) => isDayEntry(day));
}

function readCacheFile(filePath: string): Record<string, CacheEntry> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(filePath, 'utf8'));
  } catch {
    // Missing or unreadable: the logs are the source of truth, so an absent
    // cache is never an error — it only costs a full parse.
    return {};
  }

  if (
    !isRecord(parsed) ||
    parsed.format !== CACHE_FORMAT ||
    // Costs are baked in at parse time, so a release that changes the pricing
    // tables must invalidate everything rather than serve stale dollars.
    parsed.appVersion !== packageVersion() ||
    !isRecord(parsed.entries)
  ) {
    return {};
  }

  const entries: Record<string, CacheEntry> = {};
  for (const [path, entry] of Object.entries(parsed.entries)) {
    if (isCacheEntry(entry)) entries[path] = entry;
  }
  return entries;
}

export interface ParseCache {
  /** The cached parse for `filePath`, or null when absent or stale. */
  lookup: (filePath: string) => Promise<CachedParse | null>;
  record: (filePath: string, parse: CachedParse) => Promise<void>;
  /** Persist. Entries never looked up this run are dropped, so deleted logs age out. */
  save: () => void;
}

/** A cache that stores nothing, used when AITRACK_NO_CACHE is set. */
function disabledCache(): ParseCache {
  return {
    lookup: () => Promise.resolve(null),
    record: () => Promise.resolve(),
    save: () => undefined,
  };
}

/**
 * Per-file parse cache for one provider's transcripts.
 *
 * Every command re-derives usage from the whole local corpus, which is the bulk
 * of its runtime, yet transcript files are append-mostly and the overwhelming
 * majority are byte-identical between runs. Entries are keyed by absolute path
 * and validated against the file's mtime and size.
 *
 * `name` gives each provider its own file: the readers run concurrently, and a
 * shared file would have them clobbering each other's writes.
 */
export function openParseCache(name: string): ParseCache {
  if (process.env.AITRACK_NO_CACHE) return disabledCache();

  const cachePath = join(CACHE_DIR, `${name}.json`);
  const previous = readCacheFile(cachePath);
  const next: Record<string, CacheEntry> = {};

  return {
    async lookup(filePath) {
      const entry = previous[filePath];
      if (!entry) return null;
      try {
        const stats = await stat(filePath);
        if (stats.mtimeMs !== entry.mtimeMs || stats.size !== entry.size) return null;
      } catch {
        return null;
      }
      next[filePath] = entry;
      return { days: new Map(Object.entries(entry.days)), keys: entry.keys };
    },

    async record(filePath, parse) {
      let stats;
      try {
        stats = await stat(filePath);
      } catch {
        // The file vanished mid-run; nothing to key the entry on.
        return;
      }
      next[filePath] = {
        mtimeMs: stats.mtimeMs,
        size: stats.size,
        days: Object.fromEntries(parse.days),
        keys: parse.keys,
      };
    },

    save() {
      const payload = JSON.stringify({
        format: CACHE_FORMAT,
        appVersion: packageVersion(),
        entries: next,
      });
      // Write-then-rename so a concurrent reader never sees a half-written
      // file. A lost race just costs the next run a full parse.
      const temporaryPath = `${cachePath}.${String(process.pid)}.tmp`;
      try {
        mkdirSync(CACHE_DIR, { recursive: true });
        writeFileSync(temporaryPath, payload, 'utf8');
        renameSync(temporaryPath, cachePath);
      } catch {
        rmSync(temporaryPath, { force: true });
        // A cache that cannot be written (read-only home, full disk) must not
        // fail the command it was only meant to speed up.
      }
    },
  };
}

File: crates/aitrack-core/src/readers/cache.rs
//! Per-file parse cache for provider transcripts.

use crate::data::types::DayMap;
use crate::env::environment_value;
use crate::paths::cache_dir;
use crate::timezone::machine_timezone;
use crate::version::package_version;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::Write;
use std::path::{Path, PathBuf};

/// Bump when the cached shape changes.
pub const CACHE_FORMAT: u32 = 1;

/// One transcript file's contribution, as cached.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CachedParse {
    pub days: DayMap,
    /// Dedup keys this file holds (for Claude's cross-file deduplication).
    pub keys: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CacheEntry {
    #[serde(rename = "mtimeMs")]
    mtime_ms: i64,
    size: u64,
    days: DayMap,
    keys: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct CacheFile {
    format: u32,
    #[serde(rename = "appVersion")]
    app_version: String,
    timezone: String,
    entries: HashMap<String, CacheEntry>,
}

pub struct ParseCache {
    cache_path: PathBuf,
    previous: HashMap<String, CacheEntry>,
    next: HashMap<String, CacheEntry>,
}

impl ParseCache {
    /// Lookup the cached parse for `file_path`, or None when absent or stale.
    pub async fn lookup(&mut self, file_path: &Path) -> anyhow::Result<Option<CachedParse>> {
        let key = file_path.to_string_lossy().to_string();
        let entry = match self.previous.get(&key) {
            Some(e) => e,
            None => return Ok(None),
        };

        let metadata = match fs::metadata(file_path) {
            Ok(m) => m,
            Err(_) => return Ok(None),
        };

        let mtime_ms = metadata
            .modified()?
            .duration_since(std::time::UNIX_EPOCH)?
            .as_millis() as i64;

        if mtime_ms != entry.mtime_ms || metadata.len() != entry.size {
            return Ok(None);
        }

        self.next.insert(key, entry.clone());
        Ok(Some(CachedParse {
            days: entry.days.clone(),
            keys: entry.keys.clone(),
        }))
    }

    /// Record a freshly parsed file.
    pub async fn record(&mut self, file_path: &Path, parse: CachedParse) -> anyhow::Result<()> {
        let metadata = match fs::metadata(file_path) {
            Ok(m) => m,
            Err(_) => return Ok(()), // File vanished mid-run
        };

        let mtime_ms = metadata
            .modified()?
            .duration_since(std::time::UNIX_EPOCH)?
            .as_millis() as i64;

        let key = file_path.to_string_lossy().to_string();
        self.next.insert(
            key,
            CacheEntry {
                mtime_ms,
                size: metadata.len(),
                days: parse.days,
                keys: parse.keys,
            },
        );
        Ok(())
    }

    /// Persist the cache. Entries never looked up this run are dropped.
    pub fn save(&self) {
        let cache_file = CacheFile {
            format: CACHE_FORMAT,
            app_version: package_version().to_string(),
            timezone: machine_timezone(),
            entries: self.next.clone(),
        };

        let payload = match serde_json::to_string(&cache_file) {
            Ok(p) => p,
            Err(_) => return,
        };

        let temp_path = self
            .cache_path
            .with_extension(format!("{}.tmp", std::process::id()));

        // Write-then-rename for atomicity
        if let Ok(mut file) = File::create(&temp_path) {
            if file.write_all(payload.as_bytes()).is_ok() {
                let _ = fs::rename(&temp_path, &self.cache_path);
            }
        }
        let _ = fs::remove_file(&temp_path);
    }
}

fn read_cache_file(cache_path: &Path) -> HashMap<String, CacheEntry> {
    let contents = match fs::read_to_string(cache_path) {
        Ok(c) => c,
        Err(_) => return HashMap::new(),
    };

    let cache_file: CacheFile = match serde_json::from_str(&contents) {
        Ok(cf) => cf,
        Err(_) => return HashMap::new(),
    };

    // Invalidate if format, version, or timezone changed
    if cache_file.format != CACHE_FORMAT
        || cache_file.app_version != package_version()
        || cache_file.timezone != machine_timezone()
    {
        return HashMap::new();
    }

    cache_file.entries
}

/// A cache that stores nothing, used when AITRACK_NO_CACHE is set.
pub struct DisabledCache;

impl DisabledCache {
    pub async fn lookup(&mut self, _file_path: &Path) -> anyhow::Result<Option<CachedParse>> {
        Ok(None)
    }

    pub async fn record(&mut self, _file_path: &Path, _parse: CachedParse) -> anyhow::Result<()> {
        Ok(())
    }

    pub fn save(&self) {}
}

/// Open the parse cache for a provider (e.g., "claude", "codex").
pub fn open_parse_cache(name: &str) -> ParseCache {
    let cache_path = cache_dir().join(format!("{}.json", name));
    let previous = read_cache_file(&cache_path);

    ParseCache {
        cache_path,
        previous,
        next: HashMap::new(),
    }
}

/// Check if caching is disabled.
pub fn is_cache_disabled() -> bool {
    environment_value("AITRACK_NO_CACHE").is_some()
}
